Source/ExecutionEngine/CustomStackSizePoolTaskScheduler.cs

```diff
@@ -0,0 +1,78 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Boogie;
+
+/// <summary>
+/// Uses a thread pool of a configurable size, with threads with a configurable stack size,
+/// to queue tasks.
+/// </summary>
+public class CustomStackSizePoolTaskScheduler : TaskScheduler, IDisposable
+{
+  private readonly int threadCount;
+  private readonly ConcurrentQueue<Task> queue = new();
+  private readonly SemaphoreSlim isWorkAvailable = new(0);
+  private readonly Thread[] threads;
+
+  public static CustomStackSizePoolTaskScheduler Create(int stackSize, int threadCount)
+  {
+    return new CustomStackSizePoolTaskScheduler(stackSize, threadCount);
+  }
+  
+  private CustomStackSizePoolTaskScheduler(int stackSize, int threadCount)
+  {
+    this.threadCount = threadCount;
+
+    threads = new Thread[this.threadCount];
+    for (int i = 0; i < threads.Length; i++)
+    {
+      threads[i] = new Thread(WorkLoop, stackSize) { IsBackground = true };
+      threads[i].Start();
+    }
+  }
+
+  protected override void QueueTask(Task task)
+  {
+    queue.Enqueue(task);
+    isWorkAvailable.Release(1);
+  }
+
+  protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued)
+  {
+    return TryExecuteTask(task);
+  }
+
+  public override int MaximumConcurrencyLevel => threadCount;
+
+  protected override IEnumerable<Task> GetScheduledTasks()
+  {
+    return queue;
+  }
+  
+  private void WorkLoop()
+  {
+    while (true)
+    {
+      var task = BlockUntilTaskIsAvailable();
+      TryExecuteTask(task);
+    }
+  }
+
+  private Task BlockUntilTaskIsAvailable()
+  {
+    isWorkAvailable.Wait();
+    queue.TryDequeue(out var task);
```

Comment:
Worth a comment saying we don't need to check the return value or if `task` is null because of the separate semaphore (too bad `ConcurrentQueue` doesn't have a plain blocking `Dequeue` method).

Reply:
Changed the implementation to use `AsyncQueue`, which has the behavior you were hoping for from `ConcurrentQueue`.